src/System.Private.ServiceModel/src/System/ServiceModel/Channels/CommunicationObject.cs

```diff
@@ -758,14 +761,46 @@ internal void ThrowPending()
 
         internal protected virtual Task OnCloseAsync(TimeSpan timeout)
         {
-            Contract.Requires(false, "OnCloseAsync needs to be implemented on derived classes");
-            return TaskHelpers.CompletedTask();
+#if !FEATURE_NETNATIVE  // Avoid Reflection in NET Native and rely on CoreCLR to catch this problem.
```

Comment:
Reviewers: after experimenting with the original approach (namely: attempt to predict which types are external and cannot override the async methods), I found it too convoluted and fragile.  Plus it dragged in more metadata in NET Native.  Therefore I have changed the approach to (1) assert in CoreCLR builds that all our product code has overridden the methods, and (2) for types that cannot override them, redirect to the APM equivalents.

I think this is simpler and more robust.  Feedback welcome.